internal/provider/providerfunction/normalise_resource_id.go

```diff
@@ -0,0 +1,48 @@
+package providerfunction
+
+import (
+	"context"
+
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/recaser"
+	"github.com/hashicorp/terraform-plugin-framework/function"
+)
+
+type NormaliseResourceIDFunction struct{}
+
+var _ function.Function = NormaliseResourceIDFunction{}
+
+func NewNormaliseResourceID() function.Function {
+	return &NormaliseResourceIDFunction{}
+}
+
+func (a NormaliseResourceIDFunction) Metadata(ctx context.Context, request function.MetadataRequest, response *function.MetadataResponse) {
+	response.Name = "normalise_resource_id"
+}
+
+func (a NormaliseResourceIDFunction) Definition(ctx context.Context, request function.DefinitionRequest, response *function.DefinitionResponse) {
+	response.Definition = function.Definition{
+		Summary:             "normalise_resource_id",
+		Description:         "Parses and normalises the casing on an Azure Resource Manager ID into the correct casing for Terraform",
+		MarkdownDescription: "Parses and normalises the casing on an Azure Resource Manager ID into the correct casing for Terraform",
+		Parameters: []function.Parameter{
+			function.StringParameter{
+				Name:                "id",
+				Description:         "Resource ID",
+				MarkdownDescription: "Resource ID",
+			},
+		},
+		Return: function.StringReturn{},
+	}
+}
+
+func (a NormaliseResourceIDFunction) Run(ctx context.Context, request function.RunRequest, response *function.RunResponse) {
+	var id string
+
+	response.Error = function.ConcatFuncErrors(request.Arguments.Get(ctx, &id))
+
+	if response.Error != nil {
+		return
+	}
+
+	response.Error = function.ConcatFuncErrors(response.Result.Set(ctx, recaser.ReCase(id)))
```

Comment:
As mentioned offline the other week, whilst I appreciate the desire to expose a function here - unfortunately exposing this function will be problematic, since:

1. The current implementation isn't necessarily thread-safe/performance optimised
2. The returned keys for segments can/will change between minor releases as new Resource IDs are supported
3. As-is this is going to be misleading (particularly with this name) - since it (intentionally) doesn't handle user specified segments - which'll almost certainly lead to complaints without the associated changes (including using `commonids.ResourceIDReference(`, amongst other things).

More than that - this function isn't intended to be used for anything other than data coming from the API responses, so this'd meaningfully change the design considerations/constraints for this function. I suspect (although I've not tested) that this'd be problematic when using another provider built atop protocol v6 too, if that provider isn't case/segment aware?

As such, whilst this is well intentioned - since this is not intended or ready for public use - unfortunately we'll need to remove this function.

---

Instead, as mentioned [in this comment](https://github.com/hashicorp/terraform-provider-azurerm/issues/1466#issuecomment-2056079124), we'll likely want to expose a Resource ID function per-type instead (which'll mean we can then validate the expected segments match what we're expecting (particularly in the case of Constant segments, etc) - and potentially provide defaults [e.g. `azurerm_resource_group_id(azurerm_default_subscription(), "my_group"))`]).

Reply:
I think we should still expose a function like this to correct casing as even if limited will still be useful for users - and we can always adjust things to improve the experiance

1. we should fix that as the ReCaser function likelt will need to be thread safe for use in the provider?
2. we can change this from a "best effort silent fail" to be more explicit and "we will fix a known resource ID and fail if we can't recognize it" and bail out with an error if don't recognize the id
3. I think thats fine? we can change the name to normalize_resource_id_case or normalize_resource_id_case_segment_names & well document it

This would also be a good stop gap until we can create a `azurerm_resource_group_id` function or data source as well serve a different set of use cases